fixes Micle Berry not increasing accuracy on the next turn

AlexOn1ine opened the pull request:
Fixes  #5149

Pawkkie commented:
I'm not sure this test is representative, as it passes for me on vanilla master without the change.

AlexOn1ine commented:
> I'm not sure this test is representative, as it passes for me on vanilla master without the change.

oh hm, I'm not sure how else a test could be written. Though I'm pretty sure this fixes the issue

Pawkkie commented:
Did some investigating, if you just hardcode the PASSES_RANDOMLY to the fixed percentages we want, the tests check the success rate appropriately.

That being said, this solution unfortunately doesn't fix the berry applying across turns. Here are my tests:
This first one, where the Berry affects the move used in the same turn, passes
```
SINGLE_BATTLE_TEST("Miscle Berry increases the accuracy of the next used move")
{
    GIVEN {
        ASSUME(gMovesInfo[MOVE_ROCK_SLIDE].accuracy == 90);
        PASSES_RANDOMLY(100, 100, RNG_ACCURACY);
        PLAYER(SPECIES_WOBBUFFET) { MaxHP(100); HP(26); Item(ITEM_MICLE_BERRY); }
        OPPONENT(SPECIES_WOBBUFFET);
    } WHEN {
        TURN { MOVE(opponent, MOVE_TACKLE); MOVE(player, MOVE_ROCK_SLIDE); }
    } SCENE {
        ANIMATION(ANIM_TYPE_MOVE, MOVE_TACKLE, opponent);
        ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_HELD_ITEM_EFFECT, player);
        ANIMATION(ANIM_TYPE_MOVE, MOVE_ROCK_SLIDE, player);
    }
}
```
This one, which is the same one you have here with the PASSES_RANDOMLY change, fails
```
SINGLE_BATTLE_TEST("Miscle Berry increases the accuracy of the next used move across turns")
{
    GIVEN {
        ASSUME(gMovesInfo[MOVE_ROCK_SLIDE].accuracy == 90);
        PASSES_RANDOMLY(100, 100, RNG_ACCURACY);
        PLAYER(SPECIES_WOBBUFFET) { MaxHP(100); HP(26); Item(ITEM_MICLE_BERRY); }
        OPPONENT(SPECIES_WOBBUFFET);
    } WHEN {
        TURN { MOVE(opponent, MOVE_TACKLE); }
        TURN { MOVE(player, MOVE_ROCK_SLIDE); }
    } SCENE {
        ANIMATION(ANIM_TYPE_MOVE, MOVE_TACKLE, opponent);
        ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_HELD_ITEM_EFFECT, player);
        ANIMATION(ANIM_TYPE_MOVE, MOVE_ROCK_SLIDE, player);
    }
}
```

AlexOn1ine commented:
> That being said, this solution unfortunately doesn't fix the berry applying across turns. Here are my tests:
> This first one, where the Berry affects the move used in the same turn, passes

I think my test is flawed but as far as I understand it should not work across turn but for the next used move.

Pawkkie commented:
That's poor phrasing on my part, it's the next move used across turns that I was talking about, that's the only case it seems to break in.

T1 Player Wobb Celebrate, opp Wobb tackle, berry procs.
T2 Player Wobb Rock Slide should have 108 accuracy from berry, can still miss.

AlexOn1ine commented:
Fixed. The problem was that the protect struct got cleared each turn, I used a BattleStruct field instead. 

I'm not sure if the protect struct getting clear during ai calcs is optimal but it is out of scope for this pr so I added a comment for now

AlexOn1ine commented:
drafting for now. I want to do it a little bit differently

AlexOn1ine commented:
> drafting for now. I want to do it a little bit differently

Ready for review. Though didn't change anything except a typo.

Pawkkie commented on test/battle/hold_effect/micle_berry.c:
```diff
@@ -64,3 +64,36 @@ SINGLE_BATTLE_TEST("Micle Berry raises the holder's accuracy by 1.2")
+    }
+}
+
+SINGLE_BATTLE_TEST("Miscle Berry increases the accuracy of the next used move")
```
There's a typo in the berry name, should be Micle. This would mean the two tests have identical names, so I'd suggest either going with the name I was using for the other test ("Micle Berry increases the accuracy of the next used move across turns")or coming up with something else that's representative :)

Amusingly that typo was also in the test I commented whoops I read the berry name wrong lol

AlexOn1ine replied:
pushed a suggestion

Pawkkie reviewed (changes requested):
Just one test typo, looks great otherwise!

AlexOn1ine commented:
I don't understand why this does not compile. It is fine on my end

AsparagusEduardo commented:
Probably something in master has an issue

AsparagusEduardo commented:
Nvm, nothing recent seems to have the issue

AlexOn1ine commented:
merging in master fixed the problem whatever it was
